compiler/rustc_ty_utils/src/consts.rs

```diff
@@ -51,8 +51,13 @@ fn destructure_const<'tcx>(
 
             for (field, field_valtree) in iter::zip(fields, branches) {
                 let field_ty = field.ty(tcx, args);
-                let field_const =
-                    ty::Const::new_value(tcx, field_valtree.to_value().valtree, field_ty);
+                let field_const = match field_valtree.kind() {
```

Comment:
you can get rid of this entire for loop actually as `field_consts` is just `branches` (we store a list of `ty::Const` where previously we used to store a list of `ValTree` hence this logic existing). no point in matching on a `ty::Const` just to recreate it exactly as it already is